### PR TITLE
#8250: Add unary div_no_nan

### PR DESCRIPTION
**Pull request by Aswinmcw**
Issue #8250 
Reference : [Link](https://github.com/tenstorrent/tt-metal/issues/8250#issue-2285550073)

Add unary div_no_nan forward op

All Post Commit Test : https://github.com/tenstorrent/tt-metal/actions/runs/9002733735

Test result:
- File : `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_div_no_nan.py`
Screenshot 2024-05-09 at 2 49 57 PM

- File : `tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_unary_div_no_nan.py`
Screenshot 2024-05-09 at 2 49 24 PM

Doc:
Screenshot 2024-05-09 at 2 56 47 PM

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/composite/composite_ops.cpp [hunk 1]
```diff
@@ -878,6 +878,23 @@ Tensor div_no_nan(
     return operation::decorate_as_composite(__func__, _div_no_nan)(input_a, input_b, output_mem_config);
 }
 
+Tensor _unary_div_no_nan(
+    const Tensor& input_a,
+    float value,
+    const MemoryConfig& output_mem_config) {
+    Tensor zero = full_like(input_a, 0.0f, output_mem_config);
+    if(value == 0)
+        return zero;
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:887]: Move the tensor creation into the if block.

Comment (Aswinmcw) [reply tt_eager/tt_dnn/op_library/composite/composite_ops.cpp]: Done

**Review by TT-BrianLiu — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

File: docs/source/ttnn/ttnn/dependencies/tt_lib.rst [hunk 2]
```diff
@@ -307,6 +307,8 @@ Tensor elementwise operations
 
 .. autofunction:: tt_lib.tensor.div_no_nan
 
+.. autofunction:: tt_lib.tensor.unary_div_no_nan
```
Comment (cfjchu) [inline docs/source/ttnn/ttnn/dependencies/tt_lib.rst:310]: the naming convention is not being followed

**Comment by VirdhatchaniKN** [comment]
> Why is this a separate op instead of expanding `div_unary` to handle NaN?

`div_unary` would have to return result as nan/inf based on the inputs. But for `div_no_nan` , the result would be zero instead of nan/inf .Hence we created it as a separate op
Reference : [Link](https://github.com/tenstorrent/tt-metal/issues/8250#issue-2285550073)

**Comment by Aswinmcw** [comment]
> Why is this a separate op instead of expanding `div_unary` to handle NaN?

@cfjchu The existing `div_unary` op needs to return Nan as output in some cases, but in the case of `div_no_nan` it should return 0 instead of nan.
We do have this as a separate op in tensorflow so we have taken reference from there

**Comment by cfjchu** [comment]
What is currently happening with `div_unary` if we divide by zero since our hardware doesn't support NaNs? What are the output values read back?

**Comment by VirdhatchaniKN** [comment]
> What is currently happening with `div_unary` if we divide by zero since our hardware doesn't support NaNs? What are the output values read back?

Nan/inf cases are not handled as expected in `div_unary` .Below screenshot is the result of `div_unary` when the denominator/constant value = 0.0 .
Screenshot 2024-05-09 at 10 15 47 AM

But in `div_no_nan` case, the expected result is zero when the constant(denominator) is zero. Hence, we  implemented div_no_nan as a separate op. Binary version of div_no_nan has been implemented and merged to main : [Link](https://github.com/tenstorrent/tt-metal/pull/8200)

File: docs/source/ttnn/ttnn/dependencies/tt_lib.rst (hunk 2 above)
Comment (Aswinmcw) [reply docs/source/ttnn/ttnn/dependencies/tt_lib.rst:310]: Hi @cfjchu We have updated the code and inserted the screenshots of the results : [Link](https://github.com/tenstorrent/tt-metal/pull/8255#issue-2285610587)

**Comment by cfjchu** [comment]
> > What is currently happening with `div_unary` if we divide by zero since our hardware doesn't support NaNs? What are the output values read back?
> 
> Nan/inf cases are not handled as expected in `div_unary` .Below screenshot is the result of `div_unary` when the denominator/constant value = 0.0 . Screenshot 2024-05-09 at 10 15 47 AM
> 
> But in `div_no_nan` case, the expected result is zero when the constant(denominator) is zero. Hence, we implemented div_no_nan as a separate op. Binary version of div_no_nan has been implemented and merged to main : [Link](https://github.com/tenstorrent/tt-metal/pull/8200)

I would've expected the TT-result from this screen-shot to snap to max/min saturation bounds like the first element in this result tensor or actually emit NaN. Do we have an issue tracking this divergence from torch? I'll approve this PR but we should follow-up.

**Review by cfjchu — APPROVED** [review]

**Pull request merged**